packages/eslint-plugin/src/rules/no-state-access.ts

```diff
@@ -0,0 +1,112 @@
+import * as ts from "@typescript-eslint/eslint-plugin";
+import * as tsPluginUtil from "@typescript-eslint/eslint-plugin/dist/util";
+import {
+  ParserServices,
+  TSESTree,
+} from "@typescript-eslint/experimental-utils";
+import { AST_NODE_TYPES, TSNode } from "@typescript-eslint/typescript-estree";
+import { TSRuleContext, TSRuleModule } from "../types/rules";
+import { matchModel } from "../utils/matchModel";
+
+const rule: TSRuleModule = {
+  meta: {
+    docs: {
+      description:
+        "Requires that the state variable is only accessed inside a component or action",
+      category: "Possible Errors",
+      recommended: true,
+    },
+    messages: {
+      accessingState:
+        "Trying to access the state variable. This is empty outside components or functions.",
+    },
+    schema: [],
+    type: "problem",
+  },
+  create(context: TSRuleContext) {
+    const parserServices: ParserServices = tsPluginUtil.getParserServices(
+      context,
+    );
+    const checker = parserServices.program!.getTypeChecker();
+    let importsStateFromModel: boolean = false;
+    let stateImportSymbol: ts.Symbol;
+    const insideFunction: number[] = [];
+    let insideImportDeclaration: boolean = false;
+
+    return {
+      FunctionDeclaration(node: TSESTree.FunctionDeclaration) {
+        const functionTsNode = parserServices.esTreeNodeToTSNodeMap!.get<
+          TSNode
+        >(node);
+        insideFunction.push(functionTsNode.pos);
+      },
+      FunctionExpression(node: TSESTree.FunctionExpression) {
+        const functionTsNode = parserServices.esTreeNodeToTSNodeMap!.get<
+          TSNode
+        >(node);
+        insideFunction.push(functionTsNode.pos);
+      },
+      ArrowFunctionExpression(node: TSESTree.ArrowFunctionExpression) {
+        const functionTsNode = parserServices.esTreeNodeToTSNodeMap!.get<
+          TSNode
+        >(node);
+        insideFunction.push(functionTsNode.pos);
+      },
+      "FunctionDeclaration:exit"() {
+        insideFunction.pop();
+      },
+      "FunctionExpression:exit"() {
+        insideFunction.pop();
+      },
+      "ArrowFunctionExpression:exit"() {
+        insideFunction.pop();
+      },
```

Comment:
Can probably get rid of this duplication.

Reply:
Is this betetr? Not sure it is possible to have disjunctive visitors.

Reply:
I think you can achieve the same thing with:

```js
return {
  ":function": () => { insideFunctionDepth++ },
  ":function:exit": () => { insideFunctionDepth-- },
}
```